client/src/common/diagnostic.ts

```diff
@@ -268,6 +271,16 @@ class Tabs {
 
 	public isVisible(document: TextDocument | Uri): boolean {
 		const uri = document instanceof Uri ? document : document.uri;
+		if (uri.scheme === NotebookDocumentSyncFeature.CellScheme) {
+			// Notebook cells aren't in the list of tabs, but the notebook should be.
+			return Workspace.notebookDocuments.some(notebook => {
```

Comment:
I am a little worried about that code, especially since it is mainly trigger from a context where we have the notebook itself (e.g. indirectly through the matches method). So it might be worthwhile to introduce a matchNotebookDocument method that takes more information. And then we can call isVisible directly with the notebook document uri.

Reply:
It's also used in the diagnostics pull async to verify the diagnostics are for a visible document. If this code isn't here, the diagnostics nevery actually get published.

Reply:
This code:
```javascript
            if (!this.tabs.isVisible(document)) {
                this.documentStates.unTrack(PullState.document, document);
                return;
            }
```

Reply:
@dbaeumer do you think there's a different way to handle the isVisible in the pullAsync method?

Reply:
Not for pullAsync, unless we change a lot of signatures to pass the cell document together with the notebook document around. But we could add a second optional parameter to isVisible that is the container. If we know it in the context (e.g. for example through match) we could use it. If not and it is a cell document then we can still find the notebook document.

Does that make sense?

Reply:
Yes makes sense. I'll try that.

Reply:
Actually I think the way I have match and matchCell makes more sense. matchCell knows it's looking at a cell so it asks isVisible only for the notebookDocument and not the cell itself.